synapse/handlers/presence.py

```diff
@@ -926,6 +948,36 @@ async def _handle_timeouts(self) -> None:
             )
             self.external_process_last_updated_ms.pop(process_id)
 
+        # Check if any devices for these users should be discarded.
+        new_states = []
+        for user_id in users_to_check:
+            user_devices = self.user_to_device_to_current_state.get(user_id, {})
+            # Only keep the devices that have been seen recently *unless* they're busy,
+            # which do not idle out.
+            new_user_devices = {
+                device_id: device
+                for device_id, device in user_devices.items()
+                if (device.state == PresenceState.BUSY and self._busy_presence_enabled)
+                or now - device.last_active_ts <= IDLE_TIMER
+            }
+
+            # If any device has timed out, ensure that the presence state and status msg
+            # is up-to-date.
+            if len(user_devices) != len(new_user_devices):
+                # Note that we don't need to worry about resetting other information as
+                # it will also be the most up-to-date anyway.
+                presence = _combine_device_states(new_user_devices.values())
+
+                new_states.append(
+                    self.user_to_current_state.get(
+                        user_id, UserPresenceState.default(user_id)
+                    ).copy_and_replace(state=presence)
+                )
+
+                self.user_to_device_to_current_state[user_id] = new_user_devices
+        if new_states:
+            await self._update_states(new_states)
```

Comment:
If you can, try and find a way to consolidate this into the `_update_states` call at the end of the function please. Every call to `_update_states` is a source of overhead(database and sometimes federation).

Reply:
This code all got moved into `handle_timeout()`, so resolving.